File: java/src/main/java/com/tempmail/providers/NcaoriMailProvider.java
package com.tempmail.providers;

import com.tempmail.TempMailProvider;
import com.tempmail.errors.TempMailException;
import com.tempmail.http.HttpClient;
import com.tempmail.json.SimpleJson;
import com.tempmail.json.SimpleJson.JsonArray;
import com.tempmail.json.SimpleJson.JsonObject;
import com.tempmail.json.SimpleJson.JsonValue;
import com.tempmail.models.Message;
import com.tempmail.models.MessageDetail;

import java.time.Duration;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.Random;

public class NcaoriMailProvider implements TempMailProvider {

    private static final String BASE = "https://www.nca.my.id";
    private static final DateTimeFormatter DATE_FMT = DateTimeFormatter.ISO_DATE_TIME;
    private static final String[] DOMAINS = { "ncaori.my.id", "nca.my.id" };

    private static final String[] WORDS = {
        "swift", "crystal", "storm", "frost", "shadow", "ember", "azure",
        "phantom", "silver", "iron", "crimson", "golden", "neo", "cosmic", "lunar",
        "solar", "dark", "light", "void", "flux",
    };

    private static final String[] WORDS2 = {
        "core", "leaf", "forge", "wave", "peak", "gate", "pulse",
        "blade", "shard", "drift", "hive", "node", "edge", "beacon", "nova",
        "storm", "cloud", "moon", "star", "wind",
    };

    private final HttpClient client;
    private final Random random;

    public NcaoriMailProvider() {
        this.client = new HttpClient(false, true);
        this.random = new Random();
    }

    @Override
    public String generateEmail() throws TempMailException {
        String name = WORDS[random.nextInt(WORDS.length)] + "_" + WORDS2[random.nextInt(WORDS2.length)];
        String domain = DOMAINS[random.nextInt(DOMAINS.length)];
        return name + "@" + domain;
    }

    @Override
    public List<Message> getInbox(String email) throws TempMailException {
        String respStr = client.get(BASE + "/api/emails?recipient=" + email);
        JsonObject root = SimpleJson.parse(respStr).asObject();
        JsonArray arr = root.getArray("emails");
        if (arr == null) return Collections.emptyList();

        List<Message> msgs = new ArrayList<>();
        for (JsonValue v : arr.elements()) {
            JsonObject m = v.asObject();
            String id = firstOf(m.getString("id"), "");
            String sender = firstOf(m.getString("sender"), "unknown");
            String subject = firstOf(m.getString("subject"), "(no subject)");
            LocalDateTime date = parseDate(m.getString("created_at"));
            String bodyText = m.getString("body_text");
            String bodyHtml = m.getString("body_html");

            if ((bodyText != null && !bodyText.isEmpty()) || (bodyHtml != null && !bodyHtml.isEmpty())) {
                msgs.add(new MessageDetail(id, sender, subject, date,
                    bodyText != null ? bodyText : "",
                    bodyHtml != null ? bodyHtml : "",
                    Collections.emptyList()));
            } else {
                msgs.add(new Message(id, sender, subject, date));
            }
        }
        return msgs;
    }

    @Override
    public MessageDetail readMessage(String messageId) throws TempMailException {
        throw new TempMailException("Ncaori Mail+ returns full message in getInbox(). Use getInbox() then filter by id.");
    }

    @Override
    public boolean deleteEmail(String email) throws TempMailException {
        return true;
    }

    private static String firstOf(String first, String fallback) {
        return first != null && !first.isEmpty() ? first : fallback;
    }

    private static LocalDateTime parseDate(String dateStr) {
        if (dateStr == null || dateStr.isEmpty()) return LocalDateTime.now();
        try {
            return LocalDateTime.parse(dateStr, DATE_FMT);
        } catch (Exception e) {
            return LocalDateTime.now();
        }
    }

    @Override
    public Optional<Message> waitForEmail(String email, Duration timeout, Duration interval) throws TempMailException {
        long deadline = System.currentTimeMillis() + timeout.toMillis();
        while (System.currentTimeMillis() < deadline) {
            List<Message> inbox = getInbox(email);
            if (!inbox.isEmpty()) {
                return Optional.of(inbox.get(0));
            }
            try {
                Thread.sleep(interval.toMillis());
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                return Optional.empty();
            }
        }
        return Optional.empty();
    }
}

File: php/src/Providers/NcaoriMailProvider.php
<?php

declare(strict_types=1);

namespace TempMail\Providers;

use TempMail\HttpClient;
use TempMail\Models\Message;
use TempMail\Models\MessageDetail;
use TempMail\TempMailProvider;
use TempMail\Exceptions\TempMailException;

class NcaoriMailProvider implements TempMailProvider
{
    private const BASE = 'https://www.nca.my.id';
    private const DOMAINS = ['ncaori.my.id', 'nca.my.id'];

    private const WORDS = [
        'swift', 'crystal', 'storm', 'frost', 'shadow', 'ember', 'azure',
        'phantom', 'silver', 'iron', 'crimson', 'golden', 'neo', 'cosmic', 'lunar',
        'solar', 'dark', 'light', 'void', 'flux',
    ];

    private const WORDS2 = [
        'core', 'leaf', 'forge', 'wave', 'peak', 'gate', 'pulse',
        'blade', 'shard', 'drift', 'hive', 'node', 'edge', 'beacon', 'nova',
        'storm', 'cloud', 'moon', 'star', 'wind',
    ];

    private HttpClient $http;

    public function __construct(?HttpClient $http = null)
    {
        $this->http = $http ?? new HttpClient();
    }

    public function generateEmail(): string
    {
        $name = self::WORDS[array_rand(self::WORDS)] . '_' . self::WORDS2[array_rand(self::WORDS2)];
        $domain = self::DOMAINS[array_rand(self::DOMAINS)];
        return $name . '@' . $domain;
    }

    public function getInbox(string $email): array
    {
        $resp = $this->http->get(self::BASE . '/api/emails?recipient=' . urlencode($email));
        $data = $resp['body'];
        if (!is_array($data) || !isset($data['emails']) || !is_array($data['emails'])) {
            return [];
        }
        return array_values(array_map(function (array $m): Message {
            $date = isset($m['created_at']) ? new \DateTimeImmutable($m['created_at']) : new \DateTimeImmutable();
            $id = $m['id'] ?? '';
            $sender = $m['sender'] ?? 'unknown';
            $subject = $m['subject'] ?? '(no subject)';
            if (!empty($m['body_text']) || !empty($m['body_html'])) {
                return new MessageDetail(
                    id: $id,
                    sender: $sender,
                    subject: $subject,
                    date: $date,
                    bodyText: $m['body_text'] ?? '',
                    bodyHtml: $m['body_html'] ?? '',
                );
            }
            return new Message(id: $id, sender: $sender, subject: $subject, date: $date);
        }, $data['emails']));
    }

    public function readMessage(string $email, string $messageId): MessageDetail
    {
        throw new TempMailException('Ncaori Mail+ returns full message in getInbox(). Use getInbox() then filter by id.');
    }

    public function deleteEmail(string $email): bool
    {
        return true;
    }

    public function waitForEmail(string $email, int $timeout = 60, int $interval = 5): ?Message
    {
        $deadline = time() + $timeout;
        while (time() < $deadline) {
            $inbox = $this->getInbox($email);
            if (!empty($inbox)) {
                return $inbox[0];
            }
            sleep($interval);
        }
        return null;
    }
}
